pkg/operator2/configobservation/routersecret/observe_router_secret.go

```diff
@@ -0,0 +1,91 @@
+package routersecret
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/operator2/configobservation"
+)
+
+func ObserveRouterSecret(genericlisters configobserver.Listers, recorder events.Recorder, previouslyObservedConfig map[string]interface{}) (ret map[string]interface{}, _ []error) {
+	namedCertificatesPath := []string{configobservation.OAuthServerConfigPrefix, "servingInfo", "namedCertificates"}
+	defer func() {
+		ret = configobserver.Pruned(ret, namedCertificatesPath)
+	}()
+
+	listers := genericlisters.(configobservation.Listers)
+	errs := []error{}
+
+	currentNamedCertificates, _, err := unstructured.NestedFieldCopy(previouslyObservedConfig, namedCertificatesPath...)
+	if err != nil {
+		// continue on read error from existing config in an attempt to fix it
+		errs = append(errs, err)
+	}
+
+	currentNamedCertificatesSlice := []interface{}{}
+	if currentNamedCertificates != nil {
+		currentNamedCertificatesSlice = currentNamedCertificates.([]interface{})
+	}
+
+	routerSecret, err := listers.SecretsLister.Secrets("openshift-authentication").Get("v4-0-config-system-router-certs")
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	observedNamedCertificates, err := routerSecretToSNI(routerSecret)
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	observedConfig := map[string]interface{}{}
+	if err := unstructured.SetNestedField(
+		observedConfig,
+		observedNamedCertificates,
```

Comment:
`routerSecretToSNI` gives a typed list. This is an invalid type here: it must be a `[]interface{}` if it is a slice.